### DISCUSSION TITLE
Getting started with python

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [IBM/zoau-samples#37] [Q&A] [answered]
I have dabbled with Python but am by no means an expert... 
So - after copying a Python sample and then trying to run it, I get what I bet lots of people get... 

```
bash-4.3$ python3 ds_attr.py
Traceback (most recent call last):
  File "/fultonm/ds_attr.py", line 8, in <module>
    from zoautil_py import mvscmd, datasets
ModuleNotFoundError: No module named 'zoautil_py'
bash-4.3$ type -a zoaversion
zoaversion is /usr/lpp/IBM/zpm/zoau/1.2.0.1/bin/zoaversion
zoaversion is /usr/lpp/IBM/zoautil/bin/zoaversion
```

What's my next step to make Python aware of the cool Python support in ZOAU? I bet there are many ways to do this - any suggestions on the 'best' way?

**Comment by EnsonoStevenPerva** [comment]
Looks like the ZOAU package isn't installed correctly on your test system. Looks like you need to, if you have the authority to do so, follow step 6 here: https://www.ibm.com/docs/en/zoau/1.2.0?topic=installing-configuring-zoa-utilities

Hope this helps,

Steven

**Reply by MikeFultonDev** [reply]
Thanks @EnsonoStevenPerva   
As part of the install it talks about using a PYTHONPATH or pip install approach (which hasn’t been done yet) which I think is the issue. I was wondering which of those two paths I should take (presumably you did one of those after the unpax?)

**Reply by EnsonoStevenPerva** [reply]
I am a fan of the pip install. It just feels more relatable to the way any old end user would do it. 

The other way feels way more complicated, in my opinion. I’m also quite lazy lol

**Comment by ddimatos** [comment, marked as answer]
The PIP install is the preferred way if you do have a compiler (and configured) , else the PYTHONPATH is the backup plan in case you don't have a compatible compiler. 

The PYTHONPATH alternative uses a set of pre-compiled binaries we ship in ZOAU located at `${ZOAU_HOME}/lib`. One caveat to note, if you are using anything other than Python 3.8.x with the pre-compiled binaries, you must run the `./configure.py` and your Python installation must be in a RW mount temporarily so the `configure.py` can make the necessary symbolic link for Python greater than 3.8.x. 

Also note, if you are using Python 3.10, you will want to use ZOAU v1.2.0.1 which was released last week specifically to support Python 3.10. 

There are some additional reasons you might use the pre-compiled binaries (PYTHONPATH). One would be that you are not using virtual environments and users don't have RW to the Python mount, in that case it could benefit you to let users manage their own ZOAU python with PYTHONPATH, pip will try to write into the Python mount.

Much of this could possibly be resolved with Python Wheel, maybe a discussion worth starting.

**Reply by MikeFultonDev** [reply]
i'd like to mark this as 'the answer' but I don't know how to :) ... Only @EnsonoStevenPerva 's answer is showing up as eligible - perhaps because it has more votes - which is fair.

**Comment by ddimatos** [comment]
Little test you can run, assuming the rest of ZOAU is configured.

Tell the shell where the precompiled binaries are located by using an export.
```
export PYTHONPATH=${PYTHONPATH}:${ZOAU_HOME}/lib
```

Tell the shell where to find python:
```
export PYTHON_HOME=<python home>
export PYTHON=$PYTHON_HOME/bin
export LIBPATH=$PYTHON_HOME/lib:$LIBPATH
export PATH=$PATH:$PYTHON
```

Start a python interpreter:
```
python3
```

Test the python api's from inside the interpreter:
```
>>> from zoautil_py import jobs
>>> jobs.listing("//*")
>>> [STC03801, STC03803, STC03802, STC03886, STC03889, STC03895,.....]
```
